modeshape-jcr/src/main/java/org/modeshape/jcr/cache/document/WritableSessionCache.java

```diff
@@ -144,11 +147,19 @@ public WritableSessionCache( ExecutionContext context,
         this.referrerChangesForRemovedNodes = new HashMap<NodeKey, ReferrerChanges>();
         this.binaryReferencesByNodeKey = new ConcurrentHashMap<NodeKey, Set<BinaryKey>>();
         this.txns = repositoryEnvironment.getTransactions();
+        this.nodeTypes = repositoryEnvironment.nodeTypes();
     }
 
     protected final void assertInSession( SessionNode node ) {
         assert this.changedNodes.get(node.getKey()) == node : "Node " + node.getKey() + " is not in this session";
     }
+    
+    protected NodeTypes nodeTypes() {
+        if (this.nodeTypes == null && workspaceCache().repositoryEnvironment() != null) {
+            this.nodeTypes = workspaceCache().repositoryEnvironment().nodeTypes();
+        }  
```

Comment:
As long as this method is not called during startup, then this will always return non-null. I'm questioning whether the `nodeTypes` field should be marked as `volatile` to ensure it is readable from multiple threads. I think it does not need to be marked as `volatile`: if one thread calls this method and sets the non-volatile `nodeTypes` field, another thread (shortly after that) might call the method and see a null `nodeTypes` field, and that second thread will just reset the `nodeTypes` field to the same reference. IOW, I think it is okay and I'm just looking for confirmation and a comment about the logic being idempotent.